src/merkle/tiered_smt/context.rs

```diff
@@ -0,0 +1,203 @@
+                        .into();
+                self.ordered_leaves
+                    .insert(sibling_index, [sibling].to_vec());
+                self.update_leaf_to_root(sibling_index, node)?;
+            } else {
+                let domain = Felt::new(sibling_index.depth() as u64);
+                let node =
+                    Rpo256::merge_in_domain(&[remaining, sibling.value.into()], domain).into();
+                self.tier_leaves.insert(sibling_index, sibling);
+                self.update_leaf_to_root(sibling_index, node)?;
+            }
+
+            self.insert_leaf(leaf)?;
+        } else {
+            let mut leaves = self
+                .get_ordered_leaves(&index)?
+                .ok_or(StorageError::LeafNodeWithoutDigest)?;
+            leaves.push(leaf);
+
+            // the remaining path is the only distinct section of the leaves, so we can just sort
+            // the entire key and we will achieve the same effect, without the need to recompute
+            // the key. however, we still need to convert the field elements to their canonical
+            // form.
+            //
+            // TODO this naive sort is performing multiple, unnecessary montgomery reductions that
+            // could be performed only once, if optimized.
+            leaves.sort_by_key(|l| {
+                [
+                    l.key[0].as_int(),
+                    l.key[1].as_int(),
+                    l.key[2].as_int(),
+                    l.key[3].as_int(),
+                ]
+            });
+
+            // generate a node, computing its state with the incremental hash of key/value tuples
+            // of the sorted leaves.
+            let initial = [
+                index.remaining_path(leaves[0].key).into(),
+                leaves[0].value.into(),
+            ];
+            let mut node = Rpo256::merge_in_domain(&initial, MAX_DEPTH_DOMAIN);
+            for leaf in leaves.iter().skip(1) {
+                let input = [node, index.remaining_path(leaf.key).into()];
+                node = Rpo256::merge_in_domain(&input, MAX_DEPTH_DOMAIN);
+                node = Rpo256::merge_in_domain(&[node, leaf.value.into()], MAX_DEPTH_DOMAIN);
+            }
+
+            self.ordered_leaves.insert(index, leaves);
+            self.update_leaf_to_root(index, node.into())?;
+        }
+
+        Ok(())
+    }
+
+    fn update_leaf_to_root(
+        &mut self,
+        mut index: NodeIndex,
+        mut node: Word,
+    ) -> Result<(), StorageError> {
+        self.nodes.insert(index, node);
+        self.types.insert(index, ContentType::Leaf);
+
+        while !index.is_root() {
+            let sibling = index.sibling();
+            let sibling = self.get_node(&sibling)?;
+            let state = index.build_node(node, sibling);
+            index = index.reverse();
+            node = Rpo256::merge_in_domain(&state, ZERO).into();
```

Comment:
This could be a simple `Rpo256::merge()`.

Reply:
But wouldn't we couple the code to the internals of `merge_in_domain` if we assume they are the same (which they happen to be now)?

Reply:
In my mind it is the opposite. Internal nodes are hashed without the domain - but here we are assuming that domain `ZERO` means without domain. This might change in the future (though probably won't). The basic specs are:

- Internal nodes are hashed without domain.
- Leaf nodes are hashed with domain such that the domain depends on depth.

Reply:
Done! Seems we lack a function that will allow us to `hash_elements` in a domain (if that is possible). Left a TODO note